Discussion: transfer learning

alxndrkalinin opened the pull request:
Here is initial draft for [transfer learning and multimodal learning section](https://github.com/greenelab/deep-review/issues/188#issuecomment-296204170).
Feedback for clarity, accessibility, structure, word flow, and grammar would be highly appreciated. I will have time to revise later this week. If I missed any papers, please also let me know and I will try my best to include them.

agitter commented:
I'll review this.  Can you please work on the Travis build errors?  Some of them may be due to spaces instead of tabs in `tags.tsv`.  There is also an extra space in `[ @tag:Kraus2017]`.

alxndrkalinin commented:
@agitter thanks, didn't realize those are tabs. Done!

agitter commented on sections/06_discussion.md:
```diff
@@ -198,6 +198,124 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+are heuristic and have to take into account additional hyper-parameters that
+depend on specific deep architectures, sizes of pre-training and target
+datasets, and similarity of domains. Unfortunately, not many negative results
+are presented in study publications. As rare example, DeepChem, a framework
```
I think the DeepChem discussion is too long because I already wrote about it in the ligand-based screening section.  I can help cut the text if you would like since I know method.

I'd also like to briefly remind readers of the other challenges of multi-task learning raised in that section.  We don't need to repeat them here, but multi-task learning with many tasks did not necessarily help when the tasks were not sufficiently similar, and "sufficiently similar" is hard to define.  You could leave a `TODO` here noting that I will add a few more thoughts related to chemical screening.

alxndrkalinin replied:
DeepChem discussion [shortened](https://github.com/greenelab/deep-review/pull/347/files#diff-2f7eae35d8b5136fc03e28b1f36ab9f4R385)

Also [mentioned](https://github.com/greenelab/deep-review/pull/347/files#diff-2f7eae35d8b5136fc03e28b1f36ab9f4R369) limitations on assessing "similarity"

Feel free to add more, esp for related to chemical screening.

agitter commented on sections/06_discussion.md:
```diff
@@ -198,6 +198,124 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+data. We suggest that this class of models deserve deeper investigation to
+establish best practices and determine limits for the transferability of
+features between various close related and distant learning tasks. Making
+negative results and source code public helps to accelerate progress in this
```
Data sharing likely plays an important role as well, right?  What about pre-trained models as a way to jump start transfer learning?

cgreene replied:
Advocating for sharing pre-trained models raises the potential for privacy issues. This is clearly important for the categorize disease section. I wonder if we want some single sentence around here referring the reader back to the categorize section if they're looking at data that can't be shared (i.e. some sort of individual level genetic variant data).

alxndrkalinin replied:
I briefly [touched on it and cited new Google paper on privacy-preserving transfer learning](https://github.com/greenelab/deep-review/pull/347/files#diff-2f7eae35d8b5136fc03e28b1f36ab9f4R397).
We can add more if needed.

alxndrkalinin commented:
@agitter Thanks for the feedback. I will respond to this review tomorrow.

- I also can add a paragraph on multi-task learning, since it seems to be of interest.
- I can't remember image-to-DNA transfer learning example, do you know of one?

cgreene commented on sections/06_discussion.md:
```diff
@@ -198,6 +198,124 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
-### Transfer learning/transferability of features
+### Multimodal and transfer learning
 
 * https://github.com/greenelab/deep-review/issues/139#issuecomment-268901804
```
I think it's currently discussed here:
https://github.com/greenelab/deep-review/blame/master/sections/03_categorize.md#L125

Probably the best thing to do is to mention in that paragraph that we talk about transfer learning more fully in the discussion. @alxndrkalinin - do you want to make that edit to categorize?

alxndrkalinin replied:
Sure, will do.

alxndrkalinin replied:
[done](https://github.com/greenelab/deep-review/pull/347/commits/4eb8445a72eddac8ec2cdc87494bd836ebf578e3?diff=unified#diff-0a610e6af64aa3dcfb7e61eb1dcb1da0R142)

cgreene commented on sections/06_discussion.md:
```diff
@@ -198,6 +198,124 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+number of instances or labels, often leads to poor performance of
+machine learning algorithms. When trained on such datasets, deep learning
+models are particularly prone to overfitting due to their high representational
+power. However, Transfer Learning techniques also known as Domain Adaptation
```
I don't think Transfer Learning should be caps (it's not in the title). Domain Adaptation probably shouldn't be either.

alxndrkalinin replied:
fixed

cgreene commented on sections/06_discussion.md:
```diff
@@ -198,6 +198,124 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+power. However, Transfer Learning techniques also known as Domain Adaptation
+enable transfer of extracted patterns between different datasets and even
+domains. This allows a model to take advantage on larger labeled data for
+training of a model which can  then can be reused for the problem in hand. Deep
```
Probably this should include the idea that the model is trained for a related problem & reused for the problem at hand. This is implied but it would be good to explicitly state it.

alxndrkalinin replied:
[Clarified](https://github.com/greenelab/deep-review/pull/347/files#diff-2f7eae35d8b5136fc03e28b1f36ab9f4R221)

cgreene commented on sections/06_discussion.md:
```diff
@@ -198,6 +198,124 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+layer-wise hierarchical fashion, this issue is usually addressed by fixing an
+empirically chosen number of layers that preserve generic characteristics
+of both training and target datasets. Then, the model is fine-tuned by
+re-training networks' top layers on the specific dataset in order to
```
This is an example of fine tuning if you want to mention it - your call:
https://link.springer.com/article/10.1007/s10278-016-9914-9

Quick summary:
https://github.com/greenelab/deep-review/issues/156

alxndrkalinin replied:
This was a good example, thanks. Also [used in limitations](https://github.com/greenelab/deep-review/pull/347/files#diff-2f7eae35d8b5136fc03e28b1f36ab9f4R375).

cgreene commented on sections/06_discussion.md:
```diff
@@ -198,6 +198,124 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+enhancer datasets improved cell type-specific predictions, outperforming
+state-of-the-art results.
+
+Related to transfer learning, multimodal learning framework assumes
```
missing word maybe - `a multimodal learning...` perhaps?

alxndrkalinin replied:
fixed

cgreene commented on sections/06_discussion.md:
```diff
@@ -198,6 +198,124 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+of both training and target datasets. Then, the model is fine-tuned by
+re-training networks' top layers on the specific dataset in order to
+re-learn domain-specific high level concepts. Fine-tuning on specific biological
+datasets enables more focused predictions. For example, Min et al.
```
It might be worth also mentioning Basset ( https://github.com/greenelab/deep-review/issues/16 ) by @davek44 to make clear that this type of transfer learning seems to be generally applicable. In Basset the authors demonstrate speedy fine-tuning of a general model.

alxndrkalinin replied:
[Added Basset](https://github.com/greenelab/deep-review/pull/347/files#diff-2f7eae35d8b5136fc03e28b1f36ab9f4R252)

cgreene commented on sections/06_discussion.md:
```diff
@@ -198,6 +198,124 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+feature learning from cancer datasets contained gene expression,
+DNA methylation, and miRNA expression data [@tag:Liang2015_exprs_cancer].
+This approach allowed for the capture of intrinsic relationships in
+different modalities and for better clustering performance over conventional
```
A summary sentence either comparing multimodal learning with transfer learning or providing a take home message would be nice.

alxndrkalinin replied:
I ended up writing [this short paragraph](https://github.com/greenelab/deep-review/pull/347/files#diff-2f7eae35d8b5136fc03e28b1f36ab9f4R355) before going into limitations.

cgreene commented on sections/06_discussion.md:
```diff
@@ -198,6 +198,124 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+raw DNA sequence to predict Transcription Start Site-seq (TSS-seq) outputs.
+The combined model radically improves performance over separately trained
+datatype-specific networks, suggesting that it learns the synergistic
+relationship between datasets.
```
I agree. Also tagging @ueser in case there is a specific result to highlight with FIDDLE that best highlights the improved performance.

alxndrkalinin replied:
I added more examples. Suggestions on FIDDLE are welcome too.

alxndrkalinin commented:
@cgreene thanks for the constructive feedback – I will get back to this tomorrow.

agitter commented:
@alxndrkalinin the image/sequence example I was referring to is [described in this pending pull request](https://github.com/greenelab/deep-review/pull/344/commits/f17e13f6966aba5924eca0d9faa64953be489dbe#diff-826776b8bc32e91de4c5aeca44063bbeR626).  We don't need to discuss it again here, but we can refer back to it.

@cgreene I see your point about pre-trained models and privacy.  Should we advocate for this with appropriate caveats?  My main thought is that in some cases, most notably image CNNs, it can speed up development of a task-specific model if one has something like a pre-trained ImageNet classifier to work from.  I don't know whether this is a general idea, e.g. [a model zoo of TF binding neural nets](https://github.com/kundajelab/dragonn/wiki/Model-Zoo), or only beneficial for images.

agitter commented:
I fixed a tags conflict introduced when I merged #355.  Please pull this change if you need to make further edits.

agitter commented on sections/03_categorize.md:
```diff
@@ -138,9 +138,10 @@ outperformed the other methods. Of particular interest, when the authors first
-tackles new challenges.
+This indicates a potential strength of deep methods. They can repurpose
+features across tasks, improving overall predictions as the field tackles
+new challenges. This approach is known as transfer learning and further
```
"and is further discussed".  Or shorten to "This approach is known as transfer learning (see Discussion)".

alxndrkalinin commented:
Thanks, @agitter. This is almost done, will finish today and let you know.

On Apr 26, 2017 7:38 AM, "Anthony Gitter" <notifications@github.com> wrote:

> I fixed a tags conflict introduced when I merged #355
> <https://github.com/greenelab/deep-review/pull/355>. Please pull this
> change if you need to make further edits.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenelab/deep-review/pull/347#issuecomment-297371958>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABDnMtb8PEAIE8W_IIo-orIxtaEdZzlnks5rzyzGgaJpZM4NFsle>
> .
>

agitter commented on sections/06_discussion.md:
```diff
@@ -210,6 +210,184 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
 discussing here.*
 
-### Transfer learning/transferability of features
+### Multimodal, multi-task, and transfer learning
```
Do we want only one of these hyphenated?

alxndrkalinin replied:
This is how I saw these terms used most of the times in literature. Wikipedia also uses "Multimodal" and "Multi-task". So I'd say yes.

agitter replied:
Thanks for confirming, that works for me

alxndrkalinin commented on sections/06_discussion.md:
```diff
@@ -198,6 +198,124 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
 * https://github.com/greenelab/deep-review/issues/139#issuecomment-268901804
+
+As discussed above, the fact that biomedical datasets often contain a limited
+number of instances or labels, often leads to poor performance of
```
[done](https://github.com/greenelab/deep-review/pull/347/files#diff-2f7eae35d8b5136fc03e28b1f36ab9f4R215)

alxndrkalinin commented on sections/06_discussion.md:
```diff
@@ -198,6 +198,124 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+domains. This allows a model to take advantage on larger labeled data for
+training of a model which can  then can be reused for the problem in hand. Deep
+neural networks exhibit high transferability of learnt features including when
+pre-training and target sets are distant [@tag:Yosinski2014].
```
"Distant" is typically used in TL literature to show dissimilarity of inputs. [Clarified](https://github.com/greenelab/deep-review/pull/347/files#diff-2f7eae35d8b5136fc03e28b1f36ab9f4R225).

alxndrkalinin commented on sections/06_discussion.md:
```diff
@@ -198,6 +198,124 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+k-means based methods.
+
+Deep graphical models such as DBNs are considered to be well suited for
+multimodal learning tasks since they are aimed to learn joint probability
```
done

alxndrkalinin commented on sections/06_discussion.md:
```diff
@@ -198,6 +198,124 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+specialized algorithms for training and sampling.  While it is possible
+to fine-tune these models for discriminatory tasks, they are typically
+outperformed by commonly used convolutional neural networks (CNN)
+trained with backpropagation. Multimodal learning with CNNs is usually
```
It was badly formulated. [Changed](https://github.com/greenelab/deep-review/pull/347/files#diff-2f7eae35d8b5136fc03e28b1f36ab9f4R274)

alxndrkalinin commented:
@agitter @cgreene I think this is ready for the second round.

I ended up re-writing and re-structuring some of it, adding more examples.
- clarified TL intro/definitions
- few more fine-tuning examples, incl Basset
- few more multimodal examples
- added MTL paragraph and relationship with TL and multimodal
- added limitations examples
- extended conclusions, connected to privacy concerns

I did not add "image-to-DNA transfer learning", because what's done in #344 is not an example of transfer learning, but rather architecture transfer. They didn't use a network pre-trained on images to analyze sequence data. Instead sequence data was represented in an image-like format such that their model could consume it directly as it were images. This trick is indeed interesting and worth discussing, but I don't think it belongs to TL.

agitter commented:
I see your point about the sequence-to-image approach.  That was discussed thoroughly in #344 so I agree we can leave it out here.

I'm reviewing your other changes now.

agitter commented on sections/06_discussion.md:
```diff
@@ -210,6 +210,196 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+models are particularly prone to overfitting due to their high representational
+power. However, transfer learning techniques also known as domain adaptation
+enable transfer of extracted patterns between different datasets and even
+domains. This approach can be described in 2-step process:
```
"in a 2-step"

alxndrkalinin replied:
fixed

agitter commented on sections/06_discussion.md:
```diff
@@ -210,6 +210,196 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+power. However, transfer learning techniques also known as domain adaptation
+enable transfer of extracted patterns between different datasets and even
+domains. This approach can be described in 2-step process:
+(1) training a model for the base task, and (2) reusing trained model
```
"reusing the trained"

alxndrkalinin replied:
fixed

agitter commented on sections/06_discussion.md:
```diff
@@ -210,6 +210,196 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+enable transfer of extracted patterns between different datasets and even
+domains. This approach can be described in 2-step process:
+(1) training a model for the base task, and (2) reusing trained model
+for the target problem in hand. The first step allows to take advantage of
```
Missing word?  "allows `???` to take"?

alxndrkalinin replied:
fixed

agitter commented on sections/06_discussion.md:
```diff
@@ -210,6 +210,196 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+for the target problem in hand. The first step allows to take advantage of
+larger amount of data, labels, or both, to extract better feature
+representations. Transferring learnt features in deep neural networks
+improves performance compared to random features even
```
Would "randomly initialized features" work better?

alxndrkalinin replied:
yes, thank you. fixed.

agitter commented on sections/06_discussion.md:
```diff
@@ -210,6 +210,196 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+Fine-tuning on specific biological datasets enables more focused predictions.
+The Basset package [@tag:Kelley2016_basset] for prediction of functional
+activities from DNA sequences was shown to rapidly learn and accurately predict
+on new data by leveraging a model pre-trained on available "public" data.
```
Does "public" need the quotes?

alxndrkalinin replied:
No. Fixed.

agitter commented on sections/06_discussion.md:
```diff
@@ -210,6 +210,196 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+(images or video) when jointly learnt with other modalities
+(audio or text) [@tag:Ngiam2011].
+Deep graphical models such as DBNs are considered to be well suited for
+multimodal learning tasks since they are aimed to learn a joint probability
```
I think "tasks since they learn" would be a more concise phrasing

alxndrkalinin replied:
yes, thank you

agitter commented on sections/06_discussion.md:
```diff
@@ -210,6 +210,196 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+[@tag:Angermueller2016_dl_review].
+Jha et al. [@tag:Jha2017_integrative_models] showed that an integrated training
+approach delivers better performance compared to individual networks. They
+compared a number of feed-forward architectures trained on RNA-Seq data
```
Lower case "-seq" to mach the next line

alxndrkalinin replied:
fixed

agitter commented on sections/06_discussion.md:
```diff
@@ -210,6 +210,196 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+[@tag:Dahl2014_multi_qsar @tag:Ramsundar2015_multitask_drug]
+and drug toxicity prediction
+[@tag:Mayr2016_deep_tox @tag:Hughes2016_macromol_react]. Kearnes et al.
+[@tag:Kearnes2016_admet] did a systematic comparison of single-task and
```
The Kearnes conclusions were somewhat more complex that this.  They found that adding large numbers of less-related tasks actually hurt performance sometimes, right?  It could be preferable to do MTL with a smaller number of more similar tasks.

Some of this is in Treat, so we could shorten the Kearnes discussion here.

alxndrkalinin replied:
@agitter their full results are indeed more complex in detail. On your question, see their section 3.2.3 Task relatedness on page 7:

> These results support the inclusion of as much data
and as many tasks as possible when training multitask
models, without regard to task relatedness. In
light of the seemingly contradictory result from Section
3.2.2, we suggest that multitask benefits increase
with the addition of more data up to an inflection
point where additional data does not help or even
begins to degrade performance. This inflection point
is likely to vary between datasets and will depend on
relatedness to the other tasks in the model.

Basically on subsets of more related tasks results were not better than on large number of less related tasks. I tired to briefly summarize that in
> [They confirmed that multi-task learning can improve performance over single-task models. They further showed that smaller datasets tend to benefit more from multitask
learning than larger datasets. Results emphasized that multitask effects are
highly dataset-dependent, suggesting the use of dataset-specific models to
maximize overall performance.](https://github.com/greenelab/deep-review/pull/347/files#diff-2f7eae35d8b5136fc03e28b1f36ab9f4R348)

Do you think I should reformulate this? Suggestions are welcome.

agitter replied:
Now that I re-read that quoted Kearnes paragraph, I like what you have.  Thanks for clarifying.

agitter commented on sections/06_discussion.md:
```diff
@@ -210,6 +210,196 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+
+MTL approach is complementary to multimodal and transfer learning. All three
+techniques can be used together in the same model. For example, Zhang et al.
+[@tag:Zhang2015_multitask_tl] combined deep model based transfer and multi-task
```
"model-based"

alxndrkalinin replied:
fixed

agitter commented on sections/06_discussion.md:
```diff
@@ -210,6 +210,196 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+MTL approach is complementary to multimodal and transfer learning. All three
+techniques can be used together in the same model. For example, Zhang et al.
+[@tag:Zhang2015_multitask_tl] combined deep model based transfer and multi-task
+learning for cross-domain image annotation. One could imaging extending that
```
"imagine"

alxndrkalinin replied:
fixed

agitter commented on sections/06_discussion.md:
```diff
@@ -210,6 +210,196 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+techniques can be used together in the same model. For example, Zhang et al.
+[@tag:Zhang2015_multitask_tl] combined deep model based transfer and multi-task
+learning for cross-domain image annotation. One could imaging extending that
+approach to multimodal inputs as well. Common characteristic os these methods
```
Typo in "characteristic os"

agitter commented on sections/06_discussion.md:
```diff
@@ -210,6 +210,196 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+either re-training the initial model from scratch with special pre-processing
+or fine-tuning of the whole network on radiographs with heavy data augmentation
+to avoid overfitting. Exclusively fine-tuning top layers led to much lower
+validation accuracy (81.38 vs 99.5).
```
Round to 81.4?

alxndrkalinin replied:
yes, fixed

agitter commented on sections/06_discussion.md:
```diff
@@ -210,6 +210,196 @@ influence healthcare decisions? Or, is deep learning a hypothesis generation
+the model started overfitting new data [@tag:Kelley2016_basset].
+DeepChem framework successfully improved results for
+low-data drug discovery with one-shot learning, but it was not able to achieve
+any predictive power with transfer learning on closely related testing tasks
```
I'm not sure that "closely related testing tasks from different datasets" is the correct characterization.  Their one shot learning was unsuccessful when crossing between Tox21 and SIDER.  Those are rather different types of datasets.  It may be more accurate to say something like:

> DeepChem successfully improved results for low-data drug discovery with one-shot learning for related tasks.  However, it demonstrated clear limitations to cross-task generalization across unrelated tasks, specifically nuclear receptor assays and patient adverse reactions  [@tag:AltaeTran2016_one_shot].

alxndrkalinin replied:
Thanks, your suggestion looks good – used it.

cgreene commented:
@agitter : merge when you approve - thanks!

alxndrkalinin commented:
@agitter I fixed the typos, accepted your suggestions, and replied to [your question on Kearnes et al](https://github.com/greenelab/deep-review/pull/347#discussion_r113777279).

agitter commented:
Thanks for the additional round of edits @alxndrkalinin. I'm merging this.

alxndrkalinin commented:
@agitter thanks for reviewing and proofreading!
